src/Compilers/CSharp/Portable/Compilation/SpeculativeSemanticModelWithMemberModel.cs

```diff
@@ -0,0 +1,565 @@
+
+        internal NullableWalker.SnapshotManager? ParentSnapshotManagerOpt => _parentSnapshotManagerOpt;
+
+        public override bool IsSpeculativeSemanticModel => true;
+
+        public override int OriginalPositionForSpeculation => _position;
+
+        public override CSharpSemanticModel ParentModel => _parentSemanticModel;
+
+        public override CSharpCompilation Compilation => _parentSemanticModel.Compilation;
+
+        internal override CSharpSyntaxNode Root => _memberModel.Root;
+
+        public override SyntaxTree SyntaxTree => _memberModel.SyntaxTree;
+
+        public override bool IgnoresAccessibility => _parentSemanticModel.IgnoresAccessibility;
+
+        internal sealed override SemanticModel ContainingModelOrSelf => this;
+
+        private MemberSemanticModel GetEnclosingMemberModel(int position)
+        {
+            AssertPositionAdjusted(position);
+            SyntaxNode? node = Root.FindTokenIncludingCrefAndNameAttributes(position).Parent;
+            return node is null ? _memberModel : GetEnclosingMemberModel(node);
+        }
+
+        private MemberSemanticModel GetEnclosingMemberModel(SyntaxNode node)
+        {
+            if (node.SyntaxTree != SyntaxTree)
+            {
+                return _memberModel;
+            }
+
+            var attributeOrParameter = node.FirstAncestorOrSelf<SyntaxNode>(static n => n.Kind() is SyntaxKind.Attribute or SyntaxKind.Parameter);
+
+            if (attributeOrParameter is null ||
+                attributeOrParameter == Root ||
+                attributeOrParameter.Parent is null ||
+                !Root.Span.Contains(attributeOrParameter.Span))
+            {
+                return _memberModel;
+            }
+
+            MemberSemanticModel containing = GetEnclosingMemberModel(attributeOrParameter.Parent);
+
+            switch (attributeOrParameter)
+            {
+                case AttributeSyntax attribute:
+
+                    return GetOrAddModelForAttribute(containing, attribute);
+
+                case ParameterSyntax paramDecl:
+
+                    return GetOrAddModelForParameter(node, containing, paramDecl);
+
+                default:
+                    ExceptionUtilities.UnexpectedValue(attributeOrParameter);
+                    return containing;
+            }
+        }
+
+        private MemberSemanticModel GetOrAddModelForAttribute(MemberSemanticModel containing, AttributeSyntax attribute)
+        {
+            return ImmutableInterlocked.GetOrAdd(ref _childMemberModels, attribute,
+                                                 (node, binderAndModel) => CreateModelForAttribute(binderAndModel.binder, (AttributeSyntax)node, binderAndModel.model),
+                                                 (binder: containing.GetEnclosingBinder(attribute.SpanStart), model: containing));
+        }
+
+        private MemberSemanticModel GetOrAddModelForParameter(SyntaxNode node, MemberSemanticModel containing, ParameterSyntax paramDecl)
```

Comment:
This seem very close or duplicated from `SyntaxTreeSemanticModel.GetOrAddModelForParameter`. Could we factorize?

Reply:
> Could we factorize?

I considered that, but didn't think the refactoring worth the trouble.